--- quickjs-2020-11-08/quickjs-libc.c.orig	2021-04-25 15:48:36.784881562 +0800
+++ quickjs-2020-11-08/quickjs-libc.c	2021-04-25 15:50:08.773691908 +0800
@@ -3526,6 +3526,8 @@
 #define OS_PLATFORM "darwin"
 #elif defined(EMSCRIPTEN)
 #define OS_PLATFORM "js"
+#elif defined(__RTTHREAD__)
+#define OS_PLATFORM "rt-thread"
 #else
 #define OS_PLATFORM "linux"
 #endif

 #define OS_FLAG(x) JS_PROP_INT32_DEF(#x, x, JS_PROP_CONFIGURABLE )
